ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala

```diff
@@ -60,74 +70,84 @@ class ApiPackageManagementService(
   }
 
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] = {
+    val submissionId =
+      if (request.submissionId.isEmpty)
+        SubmissionId.assertFromString(UUID.randomUUID().toString)
+      else
+        SubmissionId.assertFromString(request.submissionId)
+
+    // TODO(JM): Implement computation of maximum record time from
+    // the current configuration. I am leaving this to another PR as
+    // we need this change in other services as well and this PR is
+    // large enough. For now, we'll default to 1 minute.
+    val timeToLive = 60.seconds
+    val maxRecordTime =
+      Time.Timestamp.assertFromInstant(
+        request
+          .maximumRecordTime
+          .map(TimestampConversion.toInstant)
+          .getOrElse(timeProvider.getCurrentTime.plusNanos(timeToLive.toNanos))
+      )
+
     val resultT = for {
       dar <- DarReader { case (_, x) => Try(Archive.parseFrom(x)) }
         .readArchive(
           "package-upload",
           new ZipInputStream(new ByteArrayInputStream(request.darFile.toByteArray)))
     } yield {
-      (packagesWrite.uploadPackages(dar.all, None), dar.all.map(_.getHash))
+      packagesWrite.uploadPackages(submissionId, maxRecordTime, dar.all, None)
     }
+
     resultT.fold(
       err => Future.failed(ErrorFactories.invalidArgument(err.getMessage)),
       res =>
-        FutureConverters
-          .toScala(res._1)
-          .flatMap {
-            case UploadPackagesResult.Ok =>
-              Future.successful(UploadDarFileResponse())
-            case r @ UploadPackagesResult.Overloaded =>
-              Future.failed(ErrorFactories.resourceExhausted(r.description))
-            case r @ UploadPackagesResult.InternalError(_) =>
-              Future.failed(ErrorFactories.internal(r.reason))
-            case r @ UploadPackagesResult.InvalidPackage(_) =>
-              Future.failed(ErrorFactories.invalidArgument(r.description))
-            case r @ UploadPackagesResult.ParticipantNotAuthorized =>
-              Future.failed(ErrorFactories.permissionDenied(r.description))
-            case r @ UploadPackagesResult.NotSupported =>
-              Future.failed(ErrorFactories.unimplemented(r.description))
-          }(DE)
-          .flatMap(pollUntilPersisted(res._2, _))(DE)
+      transactionsService
+        .currentLedgerEnd()
```

Comment:
It seems possible, that a race condition can be won by an upload packages request traveling through the ledger and back through the transaction service. In that case "ledgerEndBeforeRequest" is in fact "ledgerEndAfterRequest"

Reply:
Oops, my mistake. Definitely need to query it first.